### ISSUE TITLE
How to establish a bond under dhcpcd?

### ISSUE THREAD

**Issue by Jibun-no-Kage** [rsmarples/dhcpcd#84]
How to establish a bond under dhcpcd?  Love the idea of the failback profile logic, but when I setup dhcpcd for dhcp with static failback profile, and then create a bond under networking itnerfaces, I trip over the start up sequence order, such that no interfaces actually exist on reboot.  If I let dhcpcd trigger when I run systemctl restart networking, everything works.  So clearly something is stepping on something during system start up.  Any insight or suggestions?  Is there a way to create a bond strictly under dhcpcd control?  From a configuration perspective?

**Comment by Jibun-no-Kage** [comment]
Yes and no, I hoped there was a more native or supported way to do such via dhcpcd.  But it does such one way to address the issue.  Since my goal is to get away from having to use older /etc/network/interfaces methodology, I was thinking more towards dhcpcd... since I like the idea of a fallback method.  I guess I could also go the systemd route, eh, no pun intended.

**Comment by rsmarples** [comment]
dhcpcd's job is just to configure addressing and routing on a given interface.
Setting the interface up is the job of something else.

**Comment by Jibun-no-Kage** [comment]
True, but there are many scenarios outlined via google where traditional bonding does not work well with dhcpcd.  I have experienced this corundum myself, where the typical load sequence at boot trips up between dhcpcd and the networking service.  So my idea was to figure out a way to try to address this, I happen to love the idea of the static failback feature of dhcpcd.  So, I first tried using dhcpcd with a typical bond configuration in /etc/network/interfaces configuration file.  The method worked if I ran the command sequence manually but on boot, the bond was never established.  So the idea occurred to me, that maybe not using a typical bonding configuration could be eliminated, hence my question above.  Using the exit idea is the next thing to try.

**Comment by rsmarples** [comment]
I fail to see how dhcpcd can trip up with bonding - you start it with no interfaces specified or with the `-m` flag and dhcpcd will react to interfaces arriving and departing. A bond interface by itself is nothing special.

Now there might be some extra delay when dhcpcd spots the bond interface and starts up on it, but if the OS is half decent it will report no carrier on it until an interface joins it with a carrier and thus there is no delay. But this is no different from a bridge interface from dhcpcd's perspective.

**Comment by Jibun-no-Kage** [comment]
There seems to be some odd interaction.  A race condition or sequence issue with how the two services interact.  For example, if I run dhcpcd -k then run systemctl restart networking, by bond comes up and the bond0 interface has an IP via dhcp, because I have denied the slave interfaces and only allowed the bond interface.  All good.

However, if I reboot the same device, with no changes, at all, the bond interface fails, and thus no dhcp results.  If I then run dhcpcd -k and then systemctl restart networking on the just rebooted device, the bond interface works, dhcp assignment happens.

I can reproduce this on buster and bullseye, I happen to be testing on a Raspberry Pi device, but I have created the same scenario on Intel/AMD based hardware, so this is not a ARM based hardware quirk IMHO.

**Comment by rsmarples** [comment]
#76 has been merged into master and may help here if DHCPv6 or IPv6 was in play.

**Comment by Jibun-no-Kage** [comment]
When I am working on a bond configuration, typically I leave it enabled as default so it starts with system start, i.e. the unit file under systemd.  But when I am explicitly testing a bond configuration I will at times stop via -k and then let it restart when I do systemctl restart networking.  This manual restart of networking is not 'that' clean, so usually I just do a graceful restart of the system to be sure the mimic a typical system start/boot.

The original goal was to let dhcpcd work with a DHCP server, and configured with a static fall back, so the bond0 interface is always a know ip assignment, in the DHCP server, using a reservation of course.  But some distros of Linux don't always handle this well, they sometimes expect the entire bond0 setup to be in the interfaces file, regardless of a static or dhcp ip assignment elsewhere.  

There does not seem to be any explicit denial that dhcpcd can't be used with a bond configuration, so that is why I was testing and trying to figure out if it is possible... it should be because a bond is just another interface from dhcpcd perspective.  If I set the 'bonded' interfaces to denyinterface in dhcpcd.conf, and only allow dhcpcd to handle bond0, this works in some cases, on some distros.

When I opend the question, I was thinking that maybe dhcpcd could own the bond configuration, feature enhancement maybe, and this would completely retire the need for using the interfaces file.  But I realized that dhcpcd is really just a dhcp client, so it should accept whatever interface if finds available, i.e. bond0.  But that does not explain why when using bonds, and the older interfaces file, and explicitly disabling dhcpcd, everything seems to work. but when using dhcpcd, the bond fails to come up successfully at times, even if configured to use dhcp configuration.  Feels like something is stepping on something or something is not handing off control (or back to something) in the network stack.

Most of my testing has been around or on Debian 10, 11 x64 and Pi OS 10 and 11 (i.e. Debian 10, 11 ARM), just for reference.

**Comment by perkelix** [comment]
@Jibun-no-Kage Is this still an issue?

**Comment by perkelix** [comment]
Does the above still apply to 10.5.2?
